pkg/component/config.go

```diff
@@ -100,9 +101,91 @@ func ExpectedConfig(cfg map[string]interface{}) (*proto.UnitExpectedConfig, erro
 		return nil, err
 	}
 
+	if err := updateDataStreamsFromSource(result); err != nil {
+		return nil, fmt.Errorf("could not dedot 'data_stream': %w", err)
+	}
+
 	return result, nil
 }
 
+// dataStreamAndSource is a generic way to represent proto mesages
+// that contain a source field and a datastream field.
+type dataStreamAndSource interface {
+	GetDataStream() *proto.DataStream
+	GetSource() *structpb.Struct
+}
+
+func deDotDataStream(raw dataStreamAndSource) (*proto.DataStream, error) {
+	ds := raw.GetDataStream()
+	if ds == nil {
+		ds = &proto.DataStream{}
+	}
+
+	tmp := struct {
+		DataStream struct {
+			Dataset   string `config:"dataset" yaml:"dataset"`
+			Type      string `config:"type" yaml:"type"`
+			Namespace string `config:"namespace" yaml:"namespace"`
+		} `config:"data_stream" yaml:"data_stream"`
+	}{}
+
+	cfg, err := config.NewConfigFrom(raw.GetSource().AsMap())
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate config from source field: %w", err)
+	}
+
+	if err := cfg.Unpack(&tmp); err != nil {
+		return nil, fmt.Errorf("cannot unpack source field into struct: %w", err)
+	}
+
+	if (ds.Dataset != tmp.DataStream.Dataset) && (ds.Dataset != "" && tmp.DataStream.Dataset != "") {
+		return nil, errors.New("duplicated key 'datastream.dataset'")
+	}
+
+	if (ds.Type != tmp.DataStream.Type) && (ds.Type != "" && tmp.DataStream.Type != "") {
+		return nil, errors.New("duplicated key 'datastream.type'")
+	}
+
+	if (ds.Namespace != tmp.DataStream.Namespace) && (ds.Namespace != "" && tmp.DataStream.Namespace != "") {
+		return nil, errors.New("duplicated key 'datastream.namespace'")
+	}
+
+	ret := &proto.DataStream{
+		Dataset:   merge(tmp.DataStream.Dataset, ds.Dataset),
```

Comment:
Actually is the `merge` helper really needed? I don't see a way for the `tmp.DataStream` fields to be empty if the `ds` ones are not... if we can always overwrite with the unpacked values, that simplifies these assignments (and some of the duplication logic above). (On the other hand, if there is a way that `DataStream.*` could be empty when `ds.*` is not, please add a comment noting how that can happen.)

Reply:
> Actually is the merge helper really needed?

I'm not completely sure, but I believe there is a case where it might be needed. The custom logs integration creates the following YAML:
```yaml
inputs:
  - id: logfile-logs-17010054-5623-4437-8079-69fada4de0b1
    name: 'Log-Input-TestFlattenedDatastreamFleetPolicy-2023-10-09T15:25:04Z'
    revision: 1
    type: logfile
    use_output: default
    meta:
      package:
        name: log
        version: 2.3.0
    data_stream:
      namespace: testflatteneddatastreamfleetpolicynamespace17719163731420661242
    package_policy_id: 17010054-5623-4437-8079-69fada4de0b1
    streams:
      - id: logfile-log.logs-17010054-5623-4437-8079-69fada4de0b1
        data_stream:
          dataset: testflatteneddatastreamfleetpolicy-dataset
        paths:
          - /tmp/TestFlattenedDatastreamFleetPolicy3982960466/001/log.log
        ignore_older: 72h
```
The `data_stream.*` fields are defined both in the `inputs` section as well as in the `streams` section. My understanding is that somewhere before `deDotDataStream` is called, some values will be copied over and they will be present only in the `DataStream` struct. Hence this need to 'merge' the contents of `DataStream` and `source`.

If there is only one value or they're equal we just use it, if they differ then the function returns an error.